xapian-letor/api/featuremanager_internal.cc

```diff
@@ -0,0 +1,194 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+#include <config.h>
+#include <xapian.h>
+
+#include "xapian-letor/featuremanager.h"
+#include "xapian-letor/featurevector.h"
+#include "xapian-letor/ranklist.h"
+#include "featuremanager_internal.h"
+
+#include <cstring>
+#include <cstdlib>
+#include <fstream>
+
+#include <map>
+#include <string>
+
+
+using namespace Xapian;
+using namespace std;
+
+
+std::string
+FeatureManager::Internal::getdid(const Document &doc) 
+{
+    string id="";
+    string data = doc.get_data();
+    string temp_id = data.substr(data.find("url=", 0), (data.find("sample=", 0) - data.find("url=", 0)));
+    id = temp_id.substr(temp_id.rfind('/') + 1, (temp_id.rfind('.') - temp_id.rfind('/') - 1));  //to parse the actual document name associated with the         documents if any
+
+    return id;
+}
+
+int
+FeatureManager::Internal::getlabel(map<string, map<string, int> > qrel2, const Document &doc, std::string & qid) 
+{
+    int label = -1;
+    string id = getdid(doc);    
+
+    map<string, map<string, int> >::iterator outerit;
+    map<string, int>::iterator innerit;
+
+    outerit = qrel2.find(qid);
+    if (outerit != qrel2.end()) {
+	innerit = outerit->second.find(id);
+	if (innerit != outerit->second.end()) {
+        label = innerit->second;
+	}
+    }
+    return label;
+}
+
+Xapian::RankList
+FeatureManager::Internal::create_rank_list(const Xapian::MSet & mset, std::string & qid) 
+{
+    Xapian::RankList rl;
+
+    for (Xapian::MSetIterator i = mset.begin(); i != mset.end(); ++i) {
+        
+        Xapian::Document doc = i.get_document();
+        
+        // Here a weight vector can be created in future for different weights of the document 
+        // like BM25, LM etc. 
+        double weight = i.get_weight();
+
+        map<int,double> fVals = transform(doc, weight);
+        string did = getdid(doc);
+        int label = getlabel(qrel, doc, qid);
+
+        if(label!=-1) {
+            Xapian::FeatureVector fv = create_feature_vector(fVals, label, did);
+            rl.set_qid(qid);
+            rl.add_feature_vector(fv);
+        }
+    }
+    rl.normalise();
+    return rl;
+}
+
+Xapian::FeatureVector
+FeatureManager::Internal::create_feature_vector(map<int,double> fvals, int &label, std::string & did) 
+{
+    Xapian::FeatureVector fv;
+    fv.set_did(did);
+    fv.set_label(label);
+    fv.set_fvals(fvals);
+
+    return fv;
+}
+
+map<string, map<string,int> >
+FeatureManager::Internal::load_relevance(const std::string & qrel_file) 
+{
+    typedef map<string, int> Map1;      //docid and relevance judjement 0/1
+    typedef map<string, Map1> Map2;     // qid and map1
```

Comment:
In case you haven't already, let's tag this as a bad idea and fix it up later. We should avoid using typedefs to declare an internal variable that has to match the return type that's declared without typedefs :-)